src/coreWebVitals.ts

```diff
@@ -0,0 +1,121 @@
+import type { ReportHandler } from 'web-vitals';
+import { getCLS, getFCP, getFID, getLCP, getTTFB } from 'web-vitals';
+
+type CoreWebVitalsPayload = {
+	page_view_id: string | null;
+	browser_id: string | null;
+	fid: null | number;
+	cls: null | number;
+	lcp: null | number;
+	fcp: null | number;
+	ttfb: null | number;
+};
+
+const coreWebVitalsPayload: CoreWebVitalsPayload = {
+	browser_id: null,
+	page_view_id: null,
+	fid: null,
+	cls: null,
+	lcp: null,
+	fcp: null,
+	ttfb: null,
+};
+
+// By default, sample 1% of page views
+const userInSample = Math.random() < 1 / 100;
+
+let shouldSendMetrics = window.location.hash === '#forceSendMetrics';
+// unless we are forcing sending metrics for this page view
+// via initialisation or calling forceSendMetrics()
+export const forceSendMetrics = (): void => {
+	shouldSendMetrics = true;
+};
+
+const roundWithDecimals = (value: number, precision = 6): number => {
+	const power = Math.pow(10, precision);
+	return Math.round(value * power) / power;
+};
+
+const sendData = (): boolean => {
+	if (
+		!(userInSample || shouldSendMetrics) ||
+		coreWebVitalsPayload.fcp === null
+	)
+		return false;
```

Comment:
This might be out of scope of the pr but could we do this even earlier, at the top of `coreWebVitals`?
No point in registering the web-vitals callbacks if we aren't going to end up sending the results to the lake, right?

Reply:
You have to register the callbacks if you want to be able to asynchronously force metrics to be sent. Otherwise you’ll only get partial data, and TTFB and FCP are likely to be missing.